java-checks/src/main/java/org/sonar/java/checks/AssertionsInTestsCheck.java

```diff
@@ -0,0 +1,121 @@
+@SqaleConstantRemediation("10min")
+public class AssertionsInTestsCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String GENERIC_ASSERT = "org.fest.assertions.GenericAssert";
+  private static final MethodInvocationMatcher FEST_AS_METHOD = MethodInvocationMatcher.create()
+    .typeDefinition(GENERIC_ASSERT).name("as").withNoParameterConstraint();
+  private static final MethodInvocationMatcher FEST_DESCRIBED_AS_METHOD = MethodInvocationMatcher.create()
+    .typeDefinition(GENERIC_ASSERT).name("describedAs").withNoParameterConstraint();
+  private static final MethodInvocationMatcher FEST_OVERRIDE_ERROR_METHOD = MethodInvocationMatcher.create()
+    .typeDefinition(GENERIC_ASSERT).name("overridingErrorMessage").withNoParameterConstraint();
+  private static final ArrayList<MethodInvocationMatcher> ASSERTION_INVOCATION_MATCHERS = Lists.newArrayList(
+    MethodInvocationMatcher.create().typeDefinition("org.junit.Assert").name(NameCriteria.startsWith("assert")).withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition("org.junit.Assert").name("fail").withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition("junit.framework.Assert").name(NameCriteria.startsWith("assert")).withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition("junit.framework.Assert").name(NameCriteria.startsWith("fail")).withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition("org.fest.assertions.Fail").name(NameCriteria.startsWith("fail")).withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.subtypeOf(GENERIC_ASSERT)).name(NameCriteria.any()).withNoParameterConstraint()
+  );
+
+  private Deque<Boolean> methodContainsAssertion = new ArrayDeque<>();
+  private Deque<Boolean> inUnitTest = new ArrayDeque<>();
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    boolean isUnitTest = isUnitTest(tree);
+    inUnitTest.push(isUnitTest);
+    methodContainsAssertion.push(Boolean.FALSE);
+    super.visitMethod(tree);
+    inUnitTest.pop();
+    if (isUnitTest && !methodContainsAssertion.pop()) {
+      context.addIssue(tree, this, "Add at least one assertion to this test case.");
+    }
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree tree) {
+    super.visitMethodInvocation(tree);
+    if (inUnitTest.peek() && !methodContainsAssertion.peek() && isAssertion(tree)) {
+      methodContainsAssertion.pop();
+      methodContainsAssertion.push(Boolean.TRUE);
+    }
+  }
+
+  private boolean isAssertion(MethodInvocationTree mit) {
+    for (MethodInvocationMatcher methodInvocationMatcher : ASSERTION_INVOCATION_MATCHERS) {
```

Comment:
Please have a look at `MethodInvocationMatcherCollection` that should better suit this need.